src/stream/buffer.rs

```diff
@@ -0,0 +1,62 @@
+use std::mem;
+use std::prelude::v1::*;
+
+use {Async, Poll};
+use stream::{Stream, Fuse};
+
+/// An adaptor that buffers elements in a vector before passing them on as one item.
+///
+/// This adaptor will buffer up a list of items in a stream and pass on the
+/// vector used for buffering when a specified capacity has been reached. This is
+/// created by the `Stream::buffer` method.
+#[must_use = "streams do nothing unless polled"]
+pub struct Buffer<S>
+    where S: Stream
+{
+    capacity: usize, // TODO: Do we need this? Doesn't Vec::capacity() suffice?
+    items: Vec<<S as Stream>::Item>,
+    stream: Fuse<S>
+}
+
+pub fn new<S>(s: S, capacity: usize) -> Buffer<S>
+    where S: Stream
+{
+    Buffer {
+        capacity: capacity,
+        items: Vec::with_capacity(capacity),
+        stream: super::fuse::new(s),
+    }
+}
+
+impl<S> Stream for Buffer<S>
+    where S: Stream
+{
+    type Item = Vec<<S as Stream>::Item>;
+    type Error = <S as Stream>::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let maybe_next = try_ready!(self.stream.poll());
+
+        if let Some(item) = maybe_next {
+            // Push the item into the buffer and check whether it is
+            // full. If so, replace our buffer with a new and empty one
+            // and return the full one.
+            self.items.push(item);
+            if self.items.len() < self.capacity {
+                Ok(Async::NotReady)
```

Comment:
When NotReady is returned from poll, the Stream implementation is obliged to
notify the task once a new item may be ready. As no notification of any kind is
send there, a task will wait and wait, without calling poll ever again.

What should be done here instead: the Buffer::poll should continue calling the
pull of underlying stream until NotRead is returned. Effectively offloading the
responsibility for notification to the underlying stream.